src/gas.rs

```diff
@@ -13,11 +13,71 @@ pub fn update_call_index(opcodes: &mut elements::Opcodes, inserted_index: u32) {
 	}
 }
 
-enum InjectAction {
-	Spawn(u32),
-	Continue(u32),
-	Increment,
-	IncrementSpawn,
+/// A block of code represented by it's start position and cost.
+///
+/// The block typically starts with instructions such as `loop`, `block`, `if`, etc.
+///
+/// An example of block:
+///
+/// ```ignore
+/// loop
+///   i32.const 1
+///   get_local 0
+///   i32.sub
+///   tee_local 0
+///   br_if 0
+/// end
+/// ```
+///
+/// The start of the block is `i32.const 1`.
+///
+#[derive(Debug)]
+struct BlockEntry {
+	/// Index of the first instruction (aka `Opcode`) in the block.
+	start_pos: usize,
+	/// Sum of costs of all instructions until end of the block.
+	cost: usize,
+}
+
+struct Counter {
+	/// All blocks in the order of theirs start position.
+	blocks: Vec<BlockEntry>,
+
+	// Stack of blocks. Each element is an index to a `self.blocks` vector.
+	stack: Vec<usize>,
+}
+
+impl Counter {
+	fn new() -> Counter {
+		Counter {
+			stack: Vec::new(),
+			blocks: Vec::new(),
+		}
+	}
+
+	/// Begin a new block.
+	fn begin(&mut self, cursor: usize) {
+		let block_idx = self.blocks.len();
+		self.blocks.push(BlockEntry {
+			start_pos: cursor,
+			cost: 1,
+		});
+		self.stack.push(block_idx);
+	}
+
+	/// Seal the current block.
+	///
+	/// Sealed blocks have final cost which will not change later.
+	fn seal(&mut self) {
```

Comment:
better name would be finalize, no? Sealing is about integrity

Reply:
I think `seal` is fine. In fact, the name is borrowed from the similar domain.

But i'm also good with finalize and could change if you insist

Reply:
what similar domain?

Reply:
There is an algorithm described in the paper some paper `Simple and Efficient Construction of Static Single Assignment Form`. TLDR; This alogrithm processes some code block by block. When all predecessors became known for block this block is sealed, meaning no more predecessors could be found.

It seemed to me that there is analogy to what we are doing here.

Anyway, as I said I'm good with either.